From 4e9b9e9bdd76a393d56bb53942b98f7b9a888a4e Mon Sep 17 00:00:00 2001
From: liumingyuan <liumingyuan@allwinnertech.com>
Date: Mon, 8 Apr 2019 15:45:54 +0800
Subject: [PATCH] a2dp: Fix possible crash when accepting stream transport

Signed-off-by: liumingyuan <liumingyuan@allwinnertech.com>
---
 profiles/audio/a2dp.c | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/profiles/audio/a2dp.c b/profiles/audio/a2dp.c
index f20ac6e..f297a6c 100644
--- a/profiles/audio/a2dp.c
+++ b/profiles/audio/a2dp.c
@@ -104,6 +104,7 @@ struct a2dp_setup {
 	gboolean reconfigure;
 	gboolean start;
 	GSList *cb;
+	GIOChannel *io;
 	int ref;
 };
 
@@ -157,7 +158,10 @@ static struct a2dp_setup *setup_new(struct avdtp *session)
 static void setup_free(struct a2dp_setup *s)
 {
 	DBG("%p", s);
-
+	if(s->io) {
+		g_io_channel_shutdown(s->io, TRUE, NULL);
+		g_io_channel_unref(s->io);
+	}
 	setups = g_slist_remove(setups, s);
 	if (s->session)
 		avdtp_unref(s->session);
@@ -1462,6 +1466,9 @@ static void transport_cb(GIOChannel *io, GError *err, gpointer user_data)
 
 	g_io_channel_set_close_on_unref(io, FALSE);
 
+	g_io_channel_unref(setup->io);
+	setup->io = NULL;
+
 	setup_unref(setup);
 
 	return;
@@ -1506,12 +1513,19 @@ static void confirm_cb(GIOChannel *io, gpointer data)
 		if (!setup || !setup->stream)
 			goto drop;
 
+		if(setup->io) {
+			error("transport channel already exists");
+			goto drop;
+		}
+
 		if (!bt_io_accept(io, transport_cb, setup, NULL, &err)) {
 			error("bt_io_accept: %s", err->message);
 			g_error_free(err);
 			goto drop;
 		}
 
+		setup->io = g_io_channel_ref(io);
+
 		return;
 	}
 
-- 
1.9.1

