--- a/CMakeLists.txt	2019-08-01 19:07:46.000000000 +0800
+++ b/CMakeLists.txt	2019-11-14 15:50:54.013306980 +0800
@@ -171,7 +171,9 @@
 endif()
 
 # For building the GUI
-find_package(QGLViewer)
+if(G2O_USE_QGLVIEWER)
+  find_package(QGLViewer)
+endif()
 
 # shall we build the core apps using the library
 set(G2O_BUILD_APPS ON CACHE BOOL "Build g2o apps")
@@ -222,6 +224,7 @@
 	endif (DO_SSE_AUTODETECT)
 endmacro(DEFINE_SSE_VAR)
 
+if (CMAKE_SYSTEM_PROCESSOR MATCHES x86)
 # SSE optimizations:
 DEFINE_SSE_VAR(SSE2)
 DEFINE_SSE_VAR(SSE3)
@@ -251,6 +254,7 @@
   endif()
 endif()
 # End of of SSE* autodetect code -------
+endif()
 
 # Compiler specific options for gcc
 if(CMAKE_COMPILER_IS_GNUCXX)
