Index: midori-0.5.11/CMakeLists.txt
===================================================================
--- midori-0.5.11.orig/CMakeLists.txt
+++ midori-0.5.11/CMakeLists.txt
@@ -38,6 +38,8 @@ list(APPEND CMAKE_MODULE_PATH ${CMAKE_SO
 # Disallow building during install to avoid permission problems
 set(CMAKE_SKIP_INSTALL_ALL_DEPENDENCY 1)

+set(OPTS_LIBRARIES "${OPTS_LIBRARIES};stdc++")
+
 find_package(Vala REQUIRED)
 vala_require("0.16.0")
 set(VALAFLAGS ${VALAFLAGS}
