--- a/tools/topic_tools/CMakeLists.txt	2018-08-24 02:40:21.000000000 +0800
+++ b/tools/topic_tools/CMakeLists.txt	2019-01-07 14:11:06.848538769 +0800
@@ -40,11 +40,11 @@

 add_executable(mux src/mux.cpp)
 target_link_libraries(mux topic_tools ${catkin_LIBRARIES})
-add_dependencies(topic_tools ${${PROJECT_NAME}_EXPORTED_TARGETS})
+

 add_executable(demux src/demux.cpp)
 target_link_libraries(demux topic_tools ${catkin_LIBRARIES})
-add_dependencies(topic_tools ${${PROJECT_NAME}_EXPORTED_TARGETS})
+

 add_executable(relay src/relay.cpp)
 target_link_libraries(relay topic_tools ${catkin_LIBRARIES})
