--- a/block.c
+++ b/block.c
@@ -1024,8 +1024,8 @@ static int check_extroot(char *path)
 
 			fp = fopen(tag, "r");
 			if (!fp) {
-				ULOG_ERR("extroot: failed to read UUID from %s: %d (%s)\n",
-				         tag, errno, strerror(errno));
+				ULOG_ERR("extroot: failed to open %s: %d (%s), umount %s\n",
+				         tag, errno, strerror(errno), path);
 				return -1;
 			}
 
@@ -1034,12 +1034,15 @@ static int check_extroot(char *path)
 				         tag, errno, strerror(errno));
 			fclose(fp);
 
-			if (*uuid && !strcasecmp(uuid, pr->uuid))
-				return 0;
-
-			ULOG_ERR("extroot: UUID mismatch (root: %s, %s: %s)\n",
+			if (*uuid && !strcasecmp(uuid, pr->uuid)) {
+                ULOG_ERR("extroot: UUID match (root: %s, %s: %s)\n",
 			         pr->uuid, basename(path), uuid);
-			return -1;
+            } else {
+                ULOG_ERR("extroot: UUID mismatch (root: %s, %s: %s), but continue\n",
+			         pr->uuid, basename(path), uuid);
+            }
+
+            return 0;
 		}
 	}
 
