diff --git a/src/lib/elementary/efl_ui_win.c b/src/lib/elementary/efl_ui_win.c
index 824a7fd..e171225 100644
--- a/src/lib/elementary/efl_ui_win.c
+++ b/src/lib/elementary/efl_ui_win.c
@@ -4966,7 +4966,7 @@ _elm_win_finalize_internal(Eo *obj, Efl_Ui_Win_Data *sd, const char *name, Elm_W
              else if (!strcmp(enginelist[i], ELM_WAYLAND_SHM))
                tmp_sd.ee = ecore_evas_wayland_shm_new(NULL, parent_id, 0, 0, 0, 0, 0);
              else if (!strcmp(enginelist[i], ELM_WAYLAND_EGL))
-               tmp_sd.ee = ecore_evas_wayland_egl_new(NULL, parent_id, 0, 0, 0, 0, 0);
+               tmp_sd.ee = ecore_evas_wayland_egl_new(NULL, parent_id, 0, 0, 1, 1, 0);
              else if (!strcmp(enginelist[i], ELM_SOFTWARE_WIN32))
                tmp_sd.ee = ecore_evas_software_gdi_new(NULL, 0, 0, 1, 1);
              else if (!strcmp(enginelist[i], ELM_SOFTWARE_DDRAW))
@@ -5274,7 +5274,7 @@ _elm_win_finalize_internal(Eo *obj, Efl_Ui_Win_Data *sd, const char *name, Elm_W
              if (!strcmp(engine, ELM_WAYLAND_SHM))
                sd->pointer.ee = ecore_evas_wayland_shm_new(NULL, 0, 0, 0, 0, 0, 0);
              else if (!strcmp(engine, ELM_WAYLAND_EGL))
-               sd->pointer.ee = ecore_evas_wayland_egl_new(NULL, 0, 0, 0, 0, 0, 0);
+               sd->pointer.ee = ecore_evas_wayland_egl_new(NULL, 0, 0, 0, 1, 1, 0);
 
              pevas = ecore_evas_get(sd->pointer.ee);
 
diff --git a/src/modules/evas/engines/wayland_egl/evas_wl_main.c b/src/modules/evas/engines/wayland_egl/evas_wl_main.c
index 90f3d40..8a56df2 100644
--- a/src/modules/evas/engines/wayland_egl/evas_wl_main.c
+++ b/src/modules/evas/engines/wayland_egl/evas_wl_main.c
@@ -111,13 +111,15 @@ eng_window_new(Evas_Engine_Info_Wayland *einfo, int w, int h, Render_Engine_Swap
      }
 
    if (context == EGL_NO_CONTEXT) context = gw->egl_context;
-   if (eglMakeCurrent(gw->egl_disp, EGL_NO_SURFACE,
+   if (w && h)
+     eng_window_resurf(gw);
+   /*if (eglMakeCurrent(gw->egl_disp, EGL_NO_SURFACE,
                  EGL_NO_SURFACE, gw->egl_context) == EGL_FALSE)
      {
         ERR("eglMakeCurrent() fail. code=%#x", eglGetError());
         eng_window_free(gw);
         return NULL;
-     }
+     }*/
    vendor = glGetString(GL_VENDOR);
    renderer = glGetString(GL_RENDERER);
    version = glGetString(GL_VERSION);
@@ -163,8 +165,8 @@ eng_window_new(Evas_Engine_Info_Wayland *einfo, int w, int h, Render_Engine_Swap
         gw->gl_context->eglctxt = gw->egl_context;
         eng_window_use(gw);
      }
-   if (w && h)
-     eng_window_resurf(gw);
+   //if (w && h)
+   //  eng_window_resurf(gw);
    return gw;
 }
 
