From 16621b14c2d3a8b9ad1a64dd2f30db935e21b633 Mon Sep 17 00:00:00 2001
From: wuguanling <wuguanling@allwinnertech.com>
Date: Thu, 2 Jan 2020 19:41:06 +0800
Subject: [PATCH] swupdate: add paratermen for open() in writing raw data

---
 handlers/raw_handler.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/handlers/raw_handler.c b/handlers/raw_handler.c
index 7b6c7a7..2991edc 100644
--- a/handlers/raw_handler.c
+++ b/handlers/raw_handler.c
@@ -29,7 +29,7 @@ static int install_raw_image(struct img_type *img,
 	int ret;
 	int fdout;
 
-	fdout = open(img->device, O_RDWR);
+	fdout = open(img->device, O_RDWR | O_SYNC);
 	if (fdout < 0) {
 		TRACE("Device %s cannot be opened: %s",
 				img->device, strerror(errno));
-- 
1.9.1

