From e500f7bbcdb638d4e66d7430faf02b1f35d3ab0b Mon Sep 17 00:00:00 2001
From: xielinfei <xielinfei@allwinnertech.com>
Date: Thu, 11 Jan 2018 21:02:03 +0800
Subject: [PATCH] Fixed compile error

Signed-off-by: xielinfei <xielinfei@allwinnertech.com>
---
 pango/pangofc-shape.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pango/pangofc-shape.c b/pango/pangofc-shape.c
index 816aa8b..dccd69a 100644
--- a/pango/pangofc-shape.c
+++ b/pango/pangofc-shape.c
@@ -342,7 +342,7 @@ _pango_fc_shape (PangoFont           *font,
   hb_buffer_set_direction (hb_buffer, hb_direction);
   hb_buffer_set_script (hb_buffer, hb_glib_script_to_script (analysis->script));
   hb_buffer_set_language (hb_buffer, hb_language_from_string (pango_language_to_string (analysis->language), -1));
-#if HB_VERSION_ATLEAST(1,0,3)
+#if 0//HB_VERSION_ATLEAST(1,0,3)
   hb_buffer_set_cluster_level (hb_buffer, HB_BUFFER_CLUSTER_LEVEL_MONOTONE_CHARACTERS);
 #endif
   hb_buffer_set_flags (hb_buffer,
-- 
1.9.1

