--- a/OSS/CMakeModules/FindGStreamer.cmake
+++ b/OSS/CMakeModules/FindGStreamer.cmake
@@ -58,9 +58,9 @@ macro(FIND_GSTREAMER_COMPONENT _component_prefix _pkgconfig_name _library)
 
     string(REGEX MATCH "(.*)>=(.*)" _dummy "${_pkgconfig_name}")
     if ("${CMAKE_MATCH_2}" STREQUAL "")
-        pkg_check_modules(PC_${_component_prefix} "${_pkgconfig_name} >= ${GStreamer_FIND_VERSION}")
-    else ()
         pkg_check_modules(PC_${_component_prefix} ${_pkgconfig_name})
+    else ()
+        pkg_check_modules(PC_${_component_prefix} "${CMAKE_MATCH_1}>=${CMAKE_MATCH_2}")
     endif ()
        if (NOT ${_component_prefix}_INCLUDE_DIRS)
                set(${_component_prefix}_INCLUDE_DIRS ${PC_${_component_prefix}_INCLUDE_DIRS})
