--- a/src/client/linux/minidump_writer/linux_ptrace_dumper.cc
+++ b/src/client/linux/minidump_writer/linux_ptrace_dumper.cc
@@ -188,7 +188,7 @@ bool LinuxPtraceDumper::GetThreadInfoByIndex(size_t index, ThreadInfo* info) {
   if (info->ppid == -1 || info->tgid == -1)
     return false;

-#ifdef PTRACE_GETREGSET
+#if 0
   struct iovec io;
   info->GetGeneralPurposeRegisters(&io.iov_base, &io.iov_len);
   if (sys_ptrace(PTRACE_GETREGSET, tid, (void*)NT_PRSTATUS, (void*)&io) == -1) {
--- a/src/common/linux/elfutils-inl.h
+++ b/src/common/linux/elfutils-inl.h
@@ -52,7 +52,7 @@ const typename ElfClass::Shdr* FindElfSectionByName(
     int nsection) {
   assert(name != NULL);
   assert(sections != NULL);
-  assert(nsection > 0);
+

   int name_len = my_strlen(name);
   if (name_len == 0)
